_posts/2017-11-01-dockerising-aspnet-core-app.md

```diff
@@ -0,0 +1,78 @@
+
+
+#### ASP.NET Core Containerisation
+
+There are a few container technologies available but I am familiar with [Docker](https://www.docker.com/) so I decided to stick to it. Using Docker to containerise an ASP.NET Core application is fairly easy thanks to Docker's [documentation](https://docs.docker.com/) and to Microsoft base image [microsoft/aspnetcore:2.0](https://hub.docker.com/r/microsoft/aspnetcore/). The challenges are more around how to use it. By that I mean there are quite a few decisions to make: what should the container contain? Should the environment configuration be injected? How should configuration or sensitive data be injected? How should it integrate with the Continuous Integration process?
+<br/>
+I am showing below how I answered these questions. Bear in mind that I am by no mean an expert!
+
+##### Container content
+
+With Docker the container is actually the "thing" running while we talk about building **images**. For now my image contains the binaries and the default appsettings.json configuration that is shared between all environments. My Dockerfile looks very similar to the one on [microsoft/aspnetcore:2.0](https://hub.docker.com/r/microsoft/aspnetcore/) image documentation (Pay extra attention to the *Note on ports*).
+<br/>
+So what do I do in practice with that? First I build the code and I package it. Then I build the image and I run it.
+```
+dotnet build
+dotnet publish -o somewhere
+docker build -t myapp .
+docker run -p 1234:80 myapp
+```
+So very simple.
+
+##### Injecting configuration
+
+Very often applications needs extra bit of information in order to run properly: for instance web addresses or API key to connect to external services. So let's look at how to inject data to the container at start time.
+<br/>
+As far as I know they are two solutions out of the box. Either use environment variables. Or mount a volume. I couldn't figure out which solution is best so I took the simplest approach for my use case: the data I want to inject is a JSON file so it was simpler to just mount it.
+
+This is the C# code that reads the configuration:
+```
+public class Startup
+{
+    public Startup(IHostingEnvironment env)
+    {
+        var builder = new ConfigurationBuilder()
+            .SetBasePath(env.ContentRootPath)
+            .AddJsonFile("appsettings.json")
+            .AddJsonFile("/config/extra_configuration.json"); // the important line
+        Configuration = builder.Build();
+    }
+}
+...
+```
+And this is how I mount and run the container (assuming extra_configuration.json is located at /path/to/sensitive/data/dir).
+```
+docker run -p 1234:80 -v /path/to/sensitive/data/dir:/config myapp
+```
+
+<br/>
+
+So at this stage I am able to locally create a Docker image with my code and run it locally with different configurations. The process is quite manual so in the next post I am describing how I integrated it to my Continuous Integration process.
```

Comment:
Don't need the second 'locally'?

~I am describing~ _I will describe_

Integrated ~to~ _with_

Reply:
yeah I wasn't sure about locally...